toolchain/check/convert.cpp

```diff
@@ -0,0 +1,729 @@
+          : SemIR::Node::StructValue::Make(value.parse_node(), target.type_id,
+                                           value_id, new_block.id()));
+}
+
+// Returns whether `category` is a valid expression category to produce as a
+// result of a conversion with kind `target_kind`, or at most needs a temporary
+// to be materialized.
+static bool IsValidExpressionCategoryForConversionTarget(
+    SemIR::ExpressionCategory category, ConversionTarget::Kind target_kind) {
+  switch (target_kind) {
+    case ConversionTarget::Value:
+      return category == SemIR::ExpressionCategory::Value;
+    case ConversionTarget::ValueOrReference:
+    case ConversionTarget::Discarded:
+      return category == SemIR::ExpressionCategory::Value ||
+             category == SemIR::ExpressionCategory::DurableReference ||
+             category == SemIR::ExpressionCategory::EphemeralReference ||
+             category == SemIR::ExpressionCategory::Initializing;
+    case ConversionTarget::Initializer:
+    case ConversionTarget::FullInitializer:
+      return category == SemIR::ExpressionCategory::Initializing;
+  }
+}
+
+static auto PerformBuiltinConversion(Context& context, Parse::Node parse_node,
+                                     SemIR::NodeId value_id,
+                                     ConversionTarget target) -> SemIR::NodeId {
+  auto value = context.semantics_ir().GetNode(value_id);
+  auto value_type_id = value.type_id();
+  auto target_type_node = context.semantics_ir().GetNode(
+      context.semantics_ir().GetTypeAllowBuiltinTypes(target.type_id));
+
+  // Various forms of implicit conversion are supported as builtin conversions
+  // rather than being implemented purely in the Carbon prelude as `impl`s of
+  // `ImplicitAs`. There are a few reasons we need to perform some of these
+  // conversions as builtins:
+  //
+  // 1) Conversions from struct and tuple *literals* have special rules that
+  //    cannot be implemented by invoking `ImplicitAs`. Specifically, we must
+  //    recurse into the elements of the literal before performing
+  //    initialization in order to avoid unnecessary conversions between
+  //    expression categories.
+  // 2) (Not implemented yet) Conversion of a facet to a facet type depends on
+  //    the value of the facet, not only its type.
+  // 3) Some of these conversions are used while checking the library
+  //    definition of `ImplicitAs` itself or implementations of it.
+  //
+  // We also expect to see better performance by avoiding an `impl` lookup for
+  // common conversions.
+  //
+  // TODO: We should provide a debugging flag to turn off as many of these
+  // builtin conversions as we can so that we can test that they do the same
+  // thing as the library implementations.
+  //
+  // These builtin conversions all correspond to `final impl`s in the library,
+  // so we don't need to worry about `ImplicitAs` being specialized in any of
+  // these cases.
```

Comment:
If I understand correctly, this is what you're trying to say? Hoping to further remove confusion with the prior section of comments:

```suggestion
  // Even when these could be provided in Carbon code in the prelude and we do so
  // here for performance reasons, the prelude form would use `final impl`s. This
  // means we don't need to worry about `ImplicitAs` being specialized in any of
  // these cases.
```

Reply:
I don't think that's quite what I mean: there *will* be `final impl`s in the library for all of these cases. (`T impls ImplicitAs(U)` still needs to work, etc.) But we won't use them. I've taken another pass over this comment; I think it's a lot clearer now.